pressbooks/pressbooks
# Clone contributors information

This PR changes the PB API in order to include [the new contributor metadata](https://github.com/pressbooks/pressbooks/blob/dev/inc/class-contributors.php#L292-L365).
The changes in the schema to display authors, editors, contributors, illustrators, reviewers and translators is for:
- Include those in the cloning routine. It means, the metadata will be included in the cloned book. Although a new slug for each contributor will be created, if the target book contains that slug for a contributor (the admin user who is cloning the book) the default contributor metadata in the target book will be overwritten.
- Include additional authors and contributors in the export routine for Epub and xhtml formats

Other changes were made, which were agreed with @SteelWagstaff:
- The contributors name in the API are exposed with subfix and suffix, example: https://integrations.pressbooks.network/contributors/wp-json/pressbooks/v2/metadata. If suffix contains any letter from the roman numbers it will avoid to display the comma after the last name to be exposed, see: https://github.com/pressbooks/pressbooks/blob/3217c734b9f3d25d840c247a5ac5b16b6c1698cb/inc/admin/class-sitemap.php#L67, it may be a problem for example for a suffix: `Vic` and a full name: `Carl Thomson.` It will detect the `V` and will display in the API something like: `Carl Thomson Vic` although we want: `Carl Thomson, Vic` since `V` in this case is not a roman number .  It was changed for a fixed subset discussed, since we don't want to add a comma for those letter, see: https://github.com/pressbooks/pressbooks/pull/2369/files#diff-3e98df1fc7a6e8edfe8f3516f827ab9eba589a08edc9a7ab97981fca0b80e57eR638
- Taking into account the previous point, the [oxford_comma](https://github.com/pressbooks/pressbooks/blob/11e40fac4f04096c2ff2e308b6aa0e907d9df828/inc/utility/namespace.php#L1290) method causes some issues when we have contributors name exposed in the API with a comma, i.e.: `Carl Thomson, Vic`. It will be exposed in the API like to different contributors, since [it explode the list by commas](https://github.com/pressbooks/pressbooks/blob/11e40fac4f/inc/metadata/namespace.php#L261-L319). So in this PR the oxford_comma method and the implode one was changed by `;` semicolon, instead comma, to avoid this problem. Nevertheless, if some contributor contains a semicolon (for API version previous to this change) in the name it will cause problems and will split the name in 2 different contributors. It would not be a problem for this API version, since we are exposing and extracting contributors by certain character.
- **If a contributors is not a author of any chapter / front matter / back matter won't be displayed in the API and won't be cloned.**
- The contributor's picture will be cloned to the media of the targeted book, and the picture field will be pointed to that new image cloned.
- If the source book contains previous API version of this change, the contributors metadata won't be exposed in the API, and only the name will be taken, other contributor metadata won't be cloned.
- The ICML format was removed completely. 

### Test
- Create multiple contributors for a book with different fields. Some of them you could use comma in the name and roman numbers like: `II`, `IV` or `XI` por example (Max: `XII`).
- Add Contributors, authors, illustrators, reviewers, editors and translators to the book
- Assign multiple chapters/front/back matter to different contributors
- Clone the book.
- Make sure the new cloned book contains all the contributors presents in the Book Info section of the source book (authors, editors, illustrators, etc) and those that are authors of chapters, back matters and front matters.
- Make sure all the contributor metadata is included in the target book.
- You can test the metadata API endpoints and check the contributors in `<BOOK_URL>/wp-json/pressbooks/v2/metadata` and in the chapter's metadata URL: `<BOOK_URL>/wp-json/pressbooks/v2/chapters/<CHAPTER_ID>/metadata`
- Export the cloned book (or any another you have in your local environment) to PDF and EPUB 2 and 3 and make sure the book is exported correctly. The list of authors and contributors (contributors field in Book Info section, **not** the contributor's list in the Book Info > Contributors section) must be listed in at the beginning of the book.
